// args-any/dist/src/types/Operator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Operator = void 0;
/**
 * Supported `Operators` used when filtering lists
 * @default `Eq`
 */
var Operator;
(function (Operator) {
    /**
     * Equal to
     * @example
     * -option1 eq value1
     * -option2=5
     *  */
    Operator[Operator["Eq"] = 0] = "Eq";
    /**
     * Less than
     * @example
     * -option1<5
     * -option2 lt 2
     * */
    Operator[Operator["Lt"] = 1] = "Lt";
    /**
     * Less than or equal to
     * @example
     * -option1 le 2
     * -option2<=2
     */
    Operator[Operator["Le"] = 2] = "Le";
    /**
     * Greater than
     * @example
     * -option1 gt 1
     * -option2>3
     */
    Operator[Operator["Gt"] = 3] = "Gt";
    /**
     * Greater than or equal to
     * @example
     * -option1 ge 4
     * -option2>=5
     */
    Operator[Operator["Ge"] = 4] = "Ge";
    /**
     * Not equal to
     * @example
     * -option1!=5
     * -option2 ne 4
     */
    Operator[Operator["Ne"] = 5] = "Ne";
})(Operator = exports.Operator || (exports.Operator = {}));
